src/coreclr/jit/fginline.cpp

```diff
@@ -64,6 +58,154 @@ unsigned Compiler::fgCheckInlineDepthAndRecursion(InlineInfo* inlineInfo)
     return depth;
 }
 
+//------------------------------------------------------------------------
+// IsDisallowedRecursiveInline: Check whether 'info' is a recursive inline (of
+// 'ancestor'), and whether it should be disallowed.
+//
+// Return Value:
+//    True if the inline is recursive and should be disallowed.
+//
+bool Compiler::IsDisallowedRecursiveInline(InlineContext* ancestor, InlineInfo* info)
+{
+    // We disallow inlining the exact same instantiation.
+    if ((ancestor->GetCallee() == info->fncHandle) &&
+        (ancestor->GetRuntimeContext() == info->inlineCandidateInfo->exactContextHnd))
+    {
+        JITDUMP("Callsite is trivially recursive\n");
+        return true;
+    }
+
+    // None of the inline heuristics take into account that inlining will cause
+    // type/method loading for generic contexts. When polymorphic recursion is
+    // involved this can quickly consume a large amount of resources, so try to
+    // verify that we aren't inlining recursively with complex contexts.
+    //
+    // We currently approximate this by checking whether the IL code pointers
+    // are equal, since the method handles and context handles we have are
+    // going to be different (instantiated over different types).
+    if ((ancestor->GetCode() == info->inlineCandidateInfo->methInfo.ILCode) &&
```

Comment:
How should I detect the case? Do we need a new JIT-EE API for it?

Reply:
The IL cache should strip the instantiation and use generic definition `MethodDesc` as the key. I expect that it would work most of the time.

> Do we need a new JIT-EE API for it?

Probably. 

Note that you can also have a case where the inlining is going in a cycle that involves more than one method. `A<T> -> B<T> -> C<T> -> A<G<T>> -> B<G<T>> -> ...`. So you may need to check more than just the immediate ancestor with this plan.

Reply:
Ok, I will introduce a new JIT-EE method for it.

> The IL cache should strip the instantiation and use generic definition `MethodDesc` as the key.

I don't see this happening anywhere. `getMethodInfo` uses the exact method passed by the JIT, as far as I can see.

> Note that you can also have a case where the inlining is going in a cycle that involves more than one method. `A<T> -> B<T> -> C<T> -> A<G<T>> -> B<G<T>> -> ...`. So you may need to check more than just the immediate ancestor with this plan.

This already happens.

Reply:
I think this will still result in a new `MethodIL` instance being created, with its own `byte[] _ilBytes` that will end up pinned separately from existing ones. But in any case, that's besides the point -- I will add the exact check on the EE side.

Reply:
`InstantiatedMethodIL` does not have its own copy of the `_ilBytes`. It returns ILBytes from the generic method definition here: https://github.com/dotnet/runtime/blob/a48e73a3187d8d040b698cba7b6c5542ef3c7db8/src/coreclr/tools/Common/TypeSystem/IL/InstantiatedMethodIL.cs#L38-L41

Reply:
The recursive `GetMethodIL(method.GetTypicalMethodDefinition())` call will end up creating a new `EcmaMethodIL` instance for the generic method definition. That is going to make `_methodILCache` contain multiple `InstantiatedMethodIL` that all point to different `EcmaMethodIL` instances, for the same generic method definition, but with different IL byte arrays.